invokeai/frontend/web/src/app/store/middleware/listenerMiddleware/listeners/socketio/socketConnected.ts

```diff
@@ -1,33 +1,108 @@
 import { logger } from 'app/logging/logger';
-import { isInitializedChanged } from 'features/system/store/systemSlice';
+import { $baseUrl } from 'app/store/nanostores/baseUrl';
 import { size } from 'lodash-es';
+import { atom } from 'nanostores';
 import { api } from 'services/api';
+import { queueApi, selectQueueStatus } from 'services/api/endpoints/queue';
 import { receivedOpenAPISchema } from 'services/api/thunks/schema';
 import { socketConnected } from 'services/events/actions';
 
 import { startAppListening } from '../..';
 
 const log = logger('socketio');
 
+const $isFirstConnection = atom(true);
+
 export const addSocketConnectedEventListener = () => {
   startAppListening({
     actionCreator: socketConnected,
-    effect: (action, { dispatch, getState }) => {
+    effect: async (
+      action,
+      { dispatch, getState, cancelActiveListeners, delay }
+    ) => {
       log.debug('Connected');
 
-      const { nodeTemplates, config, system } = getState();
+      /**
+       * The rest of this listener has recovery logic for when the socket disconnects and reconnects.
+       *
+       * If a queue item completed during while disconnected, we need to reset the API state to re-
+       * fetch everything, updating the gallery and queue.
+       */
 
-      const { disabledTabs } = config;
+      if ($isFirstConnection.get()) {
+        // Bail on the recovery logic if this is the first connection - we don't need to recover anything
+        $isFirstConnection.set(false);
+        return;
+      }
 
-      if (!size(nodeTemplates.templates) && !disabledTabs.includes('nodes')) {
-        dispatch(receivedOpenAPISchema());
+      const { data: prevQueueStatusData } = selectQueueStatus(getState());
+
+      // Bail if queue was empty before - we have nothing to recover
+      if (
+        !prevQueueStatusData?.queue.in_progress &&
+        !prevQueueStatusData?.queue.pending
+      ) {
+        return;
       }
 
-      if (system.isInitialized) {
-        // only reset the query caches if this connect event is a *reconnect* event
+      // Else, we need to re-fetch the queue status to see if it has changed
+
+      if ($baseUrl.get()) {
+        // If we have a baseUrl (e.g. not localhost), we need to debounce the re-fetch to not hammer server
+        cancelActiveListeners();
+        // Add artificial jitter to the debounce
+        await delay(1000 + Math.random() * 1000);
+      }
+
+      try {
+        // Fetch the queue status again
+        const queueStatusRequest = dispatch(
+          await queueApi.endpoints.getQueueStatus.initiate(undefined, {
+            forceRefetch: true,
+          })
+        );
+
+        const nextQueueStatusData = await queueStatusRequest.unwrap();
+        queueStatusRequest.unsubscribe();
+
+        // If we haven't completed more items since the last check, bail
+        if (
+          prevQueueStatusData.queue.completed ===
+          nextQueueStatusData.queue.completed
```

Comment:
Would checking for `total` be more thorough? This will not catch if something is in progress

Reply:
Total will be the same even if generations start and complete while disconnected. 

But you do raise a good point, we should just check the whole queue status object here.

Reply:
It now compares the whole queue status object.